### PR TITLE
Override template loading mechanism (Issue 294)

### PR DESCRIPTION
**Pull request by createvibe**
This is an attempt to fulfill Issue #294, [Request] override template loading mechanism(s).

This PR introduces a way for users to extend Twig and implement their own custom template loaders.  It also allows you to override previously defined loaders.

For example, load a template from redis:

```
var Twig = require('twig'),
	redis = require('redis'),
	client = redis.createClient();

Twig.extend(function(Twig) {
	
	Twig.Templates.registerLoader('redis', function(location, params, callback, error_callback) {

		var precompiled = params.precompiled;

		if (typeof callback !== 'function') {
			throw new Twig.Error('The params.load callback is required for the redis loader method');
		}

		if (!params.async) {
			throw new Twig.Error('The "redis" loader method must be used asynchronously.');
		}

		client.get(location, function(err, data) {
			if (err) {
				if (typeof error_callback === 'function') {
					error_callback(err);
				}
				return;
			}

			if (precompiled) {
				params.data = JSON.parse(data);
			} else {
				params.data = data;
			}

			var template = new Twig.Template(params);
			callback(template);
		});

		return true;

	});

});
```

Fetch a template using the custom redis loader:

```
twig({
	method: 'redis', 
	name: 'my_template_name',
	load: function(template) {
		var output = template.render({data: 'here'});
	}
});
```

The business logic for 'fs' and 'ajax' loaders have been moved into their own files, twig.loader.fs.js and twig.loader.ajax.js, respectively.

Once you set the method for a template, all extends or includes spawning from that template should follow the same method.

Just as "path" and "url" are special to AJAX and FS methods, so is "name" special to custom methods.  However, custom methods also respect "id" if there is no name.

### CODE REVIEW

**Comment by r3wt** [comment]
:+1:

**Comment by r3wt** [comment]
@plepe how long do you estimate until this can land?

**Comment by plepe** [comment]
Sorry, been busy the previous days. The code looks very nice, it also has tests. I'd be willing to merge these changes.

@r3wt Did you test the change? Does it do what you expected?

**Comment by r3wt** [comment]
@plepe I have not tested the changes, however i will do so in the morning and report back with results. it looks like it should do what i want. i really appreciate the hardwork that @createvibe has put into this.

**Comment by createvibe** [comment]
@r3wt I'm happy to help out.  Hopefully it works for your needs.  Let me know if you need me to fix anything.

**Comment by r3wt** [comment]
@createvibe i can't seem to build this.

img

**Comment by createvibe** [comment]
@r3wt Interesting, it's building fine for me on a runnable instance. I just had to `npm install` first.

screenshot 2016-01-31 14 30 37

Looking at the list of files you are showing, it doesn't seem like it's trying to build the loaders.  Are you sure you are in the correct branch on my fork?

```
git clone https://github.com/createvibe/twig.js.git 
cd twig.js
git checkout issue_294
npm i
make clean
make
make test
```

**Comment by r3wt** [comment]
ah, i see. i didn't run `npm install`. amazing how people create projects and don't document how to compile it.

**Comment by createvibe** [comment]
Building the lib is documented a little here - https://github.com/justjohn/twig.js/wiki/Contributing

**Comment by r3wt** [comment]
@createvibe i built it and wrote up a quick example and this is what i get:

img

maybe i did something wrong, i dunno. i cloned directly from your repo

**Comment by createvibe** [comment]
@r3wt did you `git checkout issue_294` or are you directly on master?

**Comment by r3wt** [comment]
@createvibe yes i did, and it works as far as loading a template. now let me reference a template inline and see what happens.

**Comment by createvibe** [comment]
@r3wt Great!  Path lookups are a little different, but if all the templates come from your custom loader, you can possibly manipulate the path before fetching the file, inside your loader.

**Comment by r3wt** [comment]
i get this error: `TwigException: Cannot extend an inline template.`






http://jqmvc.openex.info/seed/views/index.twig

**Comment by createvibe** [comment]
For that, you need the "allowInlineIncludes" parameter set to true.

You can do this directly in the loader, if you want to make sure it's available, or you can do it by passing it as a twig() param.

If you look here, you can see how I implemented the tests: https://github.com/createvibe/twig.js/blob/issue_294/test/test.loaders.js#L8

**Comment by r3wt** [comment]
its kind of fubar. inline template function of twig doesn't supply a callback, so i have no idea how i'm supposed to return the template to twig.

**Comment by r3wt** [comment]
inline template doesn't work. twig spits out `undefined` for the filename and tries to fetch `undefined`:

img

i have checked and the loader is not being called with any parameters in this situation. all the arguments are undefined.

**Comment by createvibe** [comment]
Could you show me how you are defining the loader and how you are rendering the template?

I have a working example here, if you want to see it: https://plnkr.co/edit/5wGPDF1fD4v2i0x3m70S

**Comment by createvibe** [comment]
In response to twig passing async:false for inline includes, that's just how it works currently.  There's another thread about it (issue #61), but basically, twig needs to render expressions and logic blocks synchronously.  

You can see the code here for importFile, where it passes async:false manually: https://github.com/justjohn/twig.js/blob/master/src/twig.core.js#L1168

importFile is what {% include %} uses: https://github.com/justjohn/twig.js/blob/master/src/twig.logic.js#L650

**Comment by r3wt** [comment]
well synchronous XHR is deprecated and blocked in the browser.

**Comment by createvibe** [comment]
Does the plunkr I linked above crash your browser?

**Comment by createvibe** [comment]
I do see that it's deprecated on the main thread, but it doesn't crash anything for me (chrome, ff, safari).  I'm not sure what to do about that except try to fix the synchronous loading, which is not an easy thing...

**Comment by plepe** [comment]
I already tried to make rendering asynchronous, but it is 1. a lot of work 2. totally breaks compatibility, because render() can't return anything (it could return a promise ...) but would rather take a callback as additional parameter.

I guess for the inline templates, this happens while compiling; i haven't looked at that code yet ...

**Comment by plepe** [comment]
By the way, do you think this pull request is ready for merging anyway? (synchronous inline templates are a different issue than custom loaders)

**Comment by createvibe** [comment]
I've actually been looking into being able to return promises from logic blocks / tokens, and there's just a little bit of work to make it happen.  Problems come from existing code and extending other blocks, etc.  I will submit some test when I have time.

I have noticed a bug in the existing code that expects 'url' to be defined when it never is.  I will commit a fix for that, but as for this feature is concerned, I think it's good to go.

Let me know if you notice anything I should update.

**Comment by r3wt** [comment]
:-1:
